src/data/set/semiring.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2020 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import data.set.pointwise
+
+/-!
+# Sets as a semiring under union
+
+We define `set_semiring α`, an alias of `set α`, which we endow with `∪` as addition and `*` as
+multiplication. If `α` is a (commutative) monoid, `set_semiring α` is a (commutative) semiring.-/
+
+open function set
+open_locale pointwise
+
+variables {α β : Type*}
+
+/-- An alias for `set α`, which has a semiring structure given by `∪` as "addition" and pointwise
+  multiplication `*` as "multiplication". -/
+@[derive [inhabited, partial_order, order_bot]] def set_semiring (α : Type*) : Type* := set α
+
+/-- The identity function `set α → set_semiring α`. -/
+protected def set.up : set α ≃ set_semiring α := equiv.refl _
+
+namespace set_semiring
+
+/-- The identity function `set_semiring α → set α`. -/
+protected def down : set_semiring α ≃ set α := equiv.refl _
+
+@[simp] protected lemma down_up (s : set α) : s.up.down = s := rfl
+@[simp] protected lemma up_down (s : set_semiring α) : s.down.up = s := rfl
+
+-- TODO: These lemmas are not tagged `simp` because `set.le_eq_subset` simplifies the LHS
+lemma up_le_up {s t : set α} : s.up ≤ t.up ↔ s ⊆ t := iff.rfl
+lemma up_lt_up {s t : set α} : s.up < t.up ↔ s ⊂ t := iff.rfl
+
+@[simp] lemma down_subset_down {s t : set_semiring α} : s.down ⊆ t.down ↔ s ≤ t := iff.rfl
+@[simp] lemma down_ssubset_down {s t : set_semiring α} : s.down ⊂ t.down ↔ s < t := iff.rfl
+
+instance : add_comm_monoid (set_semiring α) :=
+{ add := λ s t, (s.down ∪ t.down).up,
+  zero := (∅ : set α).up,
+  add_assoc := union_assoc,
+  zero_add := empty_union,
+  add_zero := union_empty,
+  add_comm := union_comm }
+
+/- Since addition on `set_semiring` is commutative (it is set union), there is no need
+to also have the instance `covariant_class (set_semiring α) (set_semiring α) (swap (+)) (≤)`. -/
+instance covariant_class_add : covariant_class (set_semiring α) (set_semiring α) (+) (≤) :=
+⟨λ a b c, union_subset_union_right _⟩
+
+section has_mul
+variables [has_mul α]
+
+@[nolint check_reducibility]
```

Comment:
What's the right way to use localized instances in a non localized one?

Reply:
What's the error that the linter gives here?

Reply:
[Here](https://github.com/leanprover-community/mathlib/runs/6434383300?check_suite_focus=true):
```
/- The `check_reducibility` linter reports: -/
/- THE FOLLOWING INSTANCES MIGHT NOT REDUCE.
These instances contain one or more declarations that are not instances and are also not marked
`@[reducible]`. This means that type-class inference cannot unfold these declarations, which might mean that type-class inference cannot infer that two instances are definitionally equal. This can cause unexpected errors when this class occurs as an *argument* to a type-class problem. See note [reducible non-instances]. -/
-- data/set/semiring.lean
#check @set_semiring.non_unital_non_assoc_semiring /- This instance contains the declarations [set.has_mul], which are semireducible non-instances. -/
Error: set_semiring.non_unital_non_assoc_semiring - This instance contains the declarations [set.has_mul], which are semireducible non-instances.
#check @set_semiring.non_assoc_semiring /- This instance contains the declarations [set.mul_one_class], which are semireducible non-instances. -/
Error: set_semiring.non_assoc_semiring - This instance contains the declarations [set.mul_one_class], which are semireducible non-instances.
#check @set_semiring.canonically_ordered_comm_semiring /- This instance contains the declarations [set.comm_monoid], which are semireducible non-instances. -/
Error: set_semiring.canonically_ordered_comm_semiring - This instance contains the declarations [set.comm_monoid], which are semireducible non-instances.
```

Reply:
Add `@[reducible]` to all localized instances you need here?

Reply:
But no instance is localized here, that's the problem. `set_semiring` is a type synonym that won't be endowed with anything but pointwise operations, so there's no point in localizing the instances.

Reply:
You should duplicate the `has_mul` instance with something like:
```
instance : has_mul (set_semiring A) := {..set.has_mul}
```

Reply:
This still fails the `check_reducibility` linter.

Reply:
`by dsimp_result {..set.has_mul}` maybe?

Reply:
Or I guess just `{ mul := set.image2 (*) }` since that's not exactly a long spelling anyway.

You should only need to do it for the data instances.

Reply:
I was trying to avoid this, but if it's the only way... Is `check_reducibility` really flagging this rightly, though?

Reply:
I think it's plausible there's a legitimate here where unification gets confused, so I'd be inclined to play it safe.